tracing-mock/src/subscriber.rs

```diff
@@ -70,42 +307,374 @@ impl MockSubscriberBuilder {
-        self.expected.push_back(Expect::Exit(span));
+    /// Adds the expectation that the creation of a span will be
+    /// recorded next.
+    ///
+    /// This function accepts `Into<NewSpan>` instead of
+    /// [`ExpectedSpan`] directly, so it can be used to test
+    /// span fields and the span parent.
+    ///
+    /// The new span doesn't need to be entered for this expectation
+    /// to succeed.
+    ///
+    /// If a span is recorded that doesn't match the `ExpectedSpan`,
+    /// or if something else (such as an event) is recorded first,
+    /// then the expectation will fail.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tracing_mock::{expect, subscriber};
+    /// use tracing_subscriber::{subscribe::CollectExt, util::SubscriberInitExt, Subscribe};
+    ///
+    /// let span = expect::span()
+    ///     .at_level(tracing::Level::INFO)
+    ///     .named("the span we're testing")
+    ///     .with_field(expect::field("testing").with_value(&"yes"));
+    /// let (subscriber, handle) = subscriber::mock()
+    ///     .new_span(span)
+    ///     .run_with_handle();
+    ///
+    /// let _collect = tracing_subscriber::registry()
+    ///     .with(subscriber.with_filter(tracing_subscriber::filter::filter_fn(move |_meta| true)))
+    ///     .set_default();
+    ///
+    /// _ = tracing::info_span!("the span we're testing", testing = "yes");
+    ///
+    /// handle.assert_finished();
+    /// ```
+    ///
+    /// An event is recorded before the span is created, causing the
+    /// test to fail:
+    ///
+    /// ```should_panic
+    /// use tracing_mock::{expect, subscriber};
+    /// use tracing_subscriber::{subscribe::CollectExt, util::SubscriberInitExt, Subscribe};
+    ///
+    /// let span = expect::span()
+    ///     .at_level(tracing::Level::INFO)
+    ///     .named("the span we're testing")
+    ///     .with_field(expect::field("testing").with_value(&"yes"));
+    /// let (subscriber, handle) = subscriber::mock()
+    ///     .new_span(span)
+    ///     .run_with_handle();
+    ///
+    /// let _collect = tracing_subscriber::registry()
+    ///     .with(subscriber.with_filter(tracing_subscriber::filter::filter_fn(move |_meta| true)))
+    ///     .set_default();
+    ///
+    /// tracing::info!("an event");
+    /// _ = tracing::info_span!("the span we're testing", testing = "yes");
+    ///
+    /// handle.assert_finished();
+    /// ```
+    pub fn new_span<I>(mut self, new_span: I) -> Self
+    where
+        I: Into<NewSpan>,
+    {
+        self.expected.push_back(Expect::NewSpan(new_span.into()));
         self
     }
 
-    pub fn only(mut self) -> Self {
-        self.expected.push_back(Expect::Nothing);
+    /// Adds the expectation that entering a span matching the
+    /// [`ExpectedSpan`] will be recorded next.
+    ///
+    /// This expectation is generally accompanied by a call to
+    /// [`exit`] as well. If used together with [`only`], this
+    /// is necessary.
```

Comment:
maybe worth explaining why this is necessary? maybe something like
```suggestion
    /// This expectation is generally accompanied by a call to
    /// [`exit`], since an entered span will typically be exited. If used 
    /// together with [`only`], this is likely necessary, because the span
    /// will be dropped before the test completes (except in rare cases,
    /// such as if [`std::mem::forget`] is used).
```